### ISSUE TITLE
Unexpected behaviour in ResponseEntity<String> response bodies

### ISSUE THREAD

**Issue by kieranrimmer** [spring-projects/spring-boot#12841]
I have recently upgraded from spring-boot-starter-web 2.0.0 to 2.0.1 and have encountered some unexpected behaviour, specifically; ResponseEntity<String> objects are having escape characters added to the response bodies, which was not happening in version 2.0.0.

Here is a representative example... the RestController method below:

```
@RequestMapping(value = {"/app/status"}, method = RequestMethod.GET, produces = {"application/json"})
    public ResponseEntity<String> healthCheck() {
        return new ResponseEntity<>("{\"success\": true}", HttpStatus.OK);
    }
```

in version 2.0.0 this produces a useful JSON-style response: 
```{"success": true}```

however in version 2.0.1 it produces this:
```"{\"success\": true}"```

Many thanks for any and all guidance.

**Comment by kedar-joshi** [comment]
Can you share a working sample ? I am unable to reproduce this. Seems to work exactly same on both the versions.

**Comment by snicoll** [comment]
@kieranrimmer please share a sample that we can run rather than a code snippet.

**Comment by kieranrimmer** [comment]
Hi @snicoll @kedar-joshi I am travelling and unfortunately i cant distribute that applicatiin as is. I will send a working sample application that exhibits the behaviour as soon as I can. I thank you very much for responding so quickly

**Comment by tirumalesh123** [comment]
I've tested in the version 2.0.1 but i'm not able to reproduce the issue.It would be good if you share your code snippet.

**Comment by kieranrimmer** [comment]
Hi @kedar-joshi @tirumalesh123 @snicoll @spring-issuemaster @qerub 

I pruned down our pom.xml file piece by piece and there were a couple of different permutations I noticed that would fix the problem.  Please see the branches on my fork of gs-spring-boot at https://github.com/kieranrimmer/gs-spring-boot/branches

- the ```broken``` branch exhibits the undesirable behaviour
- the ```fixed_by_gson_omission``` branch corrects the behaviour by omitting the gson package
- the ```fixed_by_spring_boot_reversion``` branch corrects the behaviour by reverting spring-boot-starter-web to 2.0.0
- the ```fixed_by_mail_update``` branch corrects the behaviour by syncing the version of spring-boot-starter-mail with spring-boot-starter-web at 2.0.1

**Comment by bclozel** [comment]
TL;DR; your sample has mixed Spring versions in it which explains the strange behavior. Fixing your `spring-boot-starter-mail` dependency version solves the issue or even better, using Spring Boot's dependency management will protect you from problems like this in the future.

Some background first: by default, Spring Boot will configure Jackson for JSON (de)serialization, even if you've got Gson on the classpath. You can express your preference for Gson with `spring.http.converters.preferred-json-mapper=gson`.

Now I've taken a look at your sample and here's what's happening:
* for some reason, mixing versions enables the Gson HTTP message converter, which is added before the default message converters
* you're trying to write a `String` with the `"application/json"` Content-Type, which the Gson converter supports and happily serializes that as a simple JSON String, escaping it in the process (you can see the same thing in your browser by doing `JSON.stringify('"spring"')` which returns `""\"spring\"""`).

In the working case, the `StringHttpMessageConverter` comes first (the Jackson/Gson ones come later) and considers writing `String` instances for `"*/*"` media types. So this is written as a regular String.

Please consider using the `spring-boot-starter-parent` or at least importing the `spring-boot-dependencies` BOM for dependency management.

Thanks!
